src/daemon.cpp

```diff
@@ -27,6 +27,20 @@ Daemon::~Daemon() { MVPN_COUNT_DTOR(Daemon); }
 bool Daemon::activate(const Config& config) {
   m_lastConfig = config;
 
+  if (m_connected) {
+    if (supportServerSwitching(config)) {
+      logger.log() << "Already connected. Server switching supported.";
+      if (!switchServer(config)) {
+        return false;
+      }
+
+      emit connected();
+      return true;
+    }
+
+    logger.log() << "Already connected. Server switching not supported.";
+  }
+
```

Comment:
While this change is understandable, I think the whole method is now pretty confusing.

```
bool Daemon::activate(const Config& config) {
  m_lastConfig = config;

  if (m_connected) {
    if (supportServerSwitching(config)) {
      logger.log() << "Already connected. Server switching supported.";
      if (!switchServer(config)) {
        return false;
      }

      m_connectionDate = QDateTime::currentDateTime();
      emit connected();
      return true;
    }

    logger.log() << "Already connected. Server switching not supported.";
  }

  if (m_connected) {
    if (!deactivate(false)) {
      return false;
    }

    Q_ASSERT(!m_connected);
  }

  m_connected = run(Up, m_lastConfig);
  m_connectionDate = QDateTime::currentDateTime();

  logger.log() << "Status:" << m_connected;

  if (m_connected) {
    emit connected();
  }

  return m_connected;
}
```

sources of confusion for me:
* if m_connected is checked 3 times. between the first and second time it's not clear what would have caused m_connected to change
* there are two times that the connected signal can get emitted - they are both inside checks for m_connected
* if we're setting a member variable, why are we also returning it? does this add confusion about how to use the method and are we ever actually using that return value?

If i knew we had good tests to test all the possible paths and failure modes here I'd feel better, but I don't think we do. (happy to be corrected)

question for my c++ learning:
* the passed in config is set to the variable m_lastConfig. the call to supportServerSwitching then uses config. why not use m_lastConfig, which is passed around elsewhere in the method and why pass it around at all when we're calling functions on the class?

Reply:
- I agree with you, this method needs to be refactored. 
- about storing m_lastConfig at the beginning of this method, that is a bug. We should not store it before calling supportServerSwitch(). 
- about passing m_lastConfig, we can remove it from run()/serverSwitch(), yes. Follow up?

Reply:
> Follow up?
Definitely should I open an issue?

Reply:
Yes, please! In the meantime, I merge the latest commit and I close this PR. is it ok?